### PR TITLE
[SPARK-29405][SQL] Alter table / Insert statements should not change a table's ownership

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

In this change, we give preference to the original table's owner if it is not empty.

### Why are the changes needed?

When executing 'insert into/overwrite ...' DML, or 'alter table set tblproperties ...'  DDL, spark would change the ownership of the table the one who runs the spark application.

### Does this PR introduce any user-facing change?

NO

### How was this patch tested?

Compare with the behavior of Apache Hive

### CODE REVIEW

**Comment by yaooqinn** [comment]
test this please

**Comment by yaooqinn** [comment]
retest this please

**Comment by yaooqinn** [comment]
cc @wangyum

**Comment by wangyum** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 1]
```diff
@@ -573,8 +574,9 @@ private[hive] class HiveClientImpl(
     // If users explicitly alter these Hive-specific properties through ALTER TABLE DDL, we respect
     // these user-specified values.
     verifyColumnDataType(table.dataSchema)
+    val owner = Option(table.owner).filter(_.nonEmpty).getOrElse(userName)
```
[the remark this answers is not in the document]
Comment (yaooqinn) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: It sounds like a bigger deal than it is. Indeed,  I am going to add `alter table owner` syntax, which may make this kind of change back and forth

**Comment by yaooqinn** [comment]
also cc @cloud-fan

**Comment by cloud-fan** [comment]
how is INSERT TABLE fixed in this PR?

**Comment by yaooqinn** [comment]
@cloud-fan https://github.com/apache/spark/blob/a45739d97ed83dfbf8200e01b23cd875735e3a74/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala#L112 
would eventually call `alterTable` here

**Comment by cloud-fan** [comment]
can we add a test in `VersionsSuite`?

**Review by cloud-fan — APPROVED** [review]

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -27,6 +27,7 @@ import scala.collection.JavaConverters._
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
 
+import org.apache.commons.lang3.StringUtils
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Please remove this line.

**Comment by wangyum** [comment]
To clarify, this PR fixed the case of `ALTER TABLE` and `INSERT TABLE`, but we still have owner issues when creating table: [SPARK-29498](https://issues.apache.org/jira/browse/SPARK-29498).

**Review by wangyum — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by HyukjinKwon — COMMENTED** [review]
LGTM too, though.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Thank you @HyukjinKwon Fixed by https://github.com/apache/spark/pull/26160.

**Pull request closed without merging**